dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java

```diff
@@ -0,0 +1,159 @@
+import alluxio.underfs.UnderFileSystem;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Optional;
+
+/**
+ * The Dora metadata manager that orchestrates the metadata operations and provides the
+ * concurrency control.
+ *
+ * TODO(elega) Invalidating page cache synchronously causes performance issue and currently it
+ *  also lacks concurrency control. Address this problem in the future.
+ */
+public class DoraMetaManager {
+  private final DoraMetaStore mMetastore;
+  private final CacheManager mCacheManager;
+  private final PagedDoraWorker mDoraWorker;
+  private final UnderFileSystem mUfs;
+  private boolean mPopulateMetadataFingerprint =
+      Configuration.getBoolean(PropertyKey.DORA_WORKER_POPULATE_METADATA_FINGERPRINT);
+
+  /**
+   * Creates a dora meta manager.
+   * @param doraWorker the dora worker instance
+   * @param metaStore the dora meta store
+   * @param cacheManger the cache manager to manage the page cache
+   * @param ufs the associated ufs
+   */
+  public DoraMetaManager(
+      PagedDoraWorker doraWorker, DoraMetaStore metaStore, CacheManager cacheManger,
+      UnderFileSystem ufs) {
+    mMetastore = metaStore;
+    mCacheManager = cacheManger;
+    mDoraWorker = doraWorker;
+    mUfs = ufs;
+  }
+
+  /**
+   * Gets file meta from UFS.
+   * @param path the full ufs path
+   * @return the file status, or empty optional if not found
+   */
+  public Optional<FileStatus> getFromUfs(String path) throws IOException {
+    try {
+      UfsStatus status = mUfs.getStatus(path);
+      DoraMeta.FileStatus fs = mDoraWorker.buildFileStatusFromUfsStatus(status, path);
+      return Optional.ofNullable(fs);
+    } catch (FileNotFoundException e) {
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Gets file meta from UFS and loads it into metastore if exists.
+   * @param path the full ufs path
+   * @return the file status, or empty optional if not found
+   */
+  public Optional<FileStatus> loadFromUfs(String path) throws IOException {
+    Optional<FileStatus> fileStatus = getFromUfs(path);
+    fileStatus.ifPresent(status -> put(path, status));
+    return fileStatus;
+  }
+
+  /**
+   * Gets file meta from the metastore.
+   * @param path the full ufs path
+   * @return the file status, or empty optional if not found
+   */
+  public Optional<FileStatus> getFromMetaStore(String path) {
+    return mMetastore.getDoraMeta(path);
+  }
+
+  /**
+   * Puts meta of a file into the metastore, and invalidates the file data cache.
+   * @param path the full ufs path
+   * @param meta the file meta
+   */
+  public void put(String path, FileStatus meta) {
+    Optional<FileStatus> status = mMetastore.getDoraMeta(path);
+    if (status.isEmpty()
+        || status.get().getFileInfo().getFolder()
+        || status.get().getFileInfo().getLength() == 0) {
+      mMetastore.putDoraMeta(path, meta);
```

Comment:
do we put directory metadata into RocksDB as well?

Reply:
yes we do